src/main/java/com/deusgmbh/xcusatio/generator/ExcuseGenerator.java

```diff
@@ -0,0 +1,167 @@
+import com.deusgmbh.xcusatio.data.usersettings.ExcusesVibes;
+
+/**
+ * 
+ * @author Tobias.Schmidt@de.ibm.com
+ *
+ */
+public class ExcuseGenerator {
+
+    public ExcuseGenerator() {
+        super();
+    }
+
+    /**
+     * 
+     * @param context
+     * @returns the percentage indicating the approval rate to be displayed
+     */
+    public double getThumbGesture(Context context) {
+        // TODO: calculate random percantage base on context (e.g. more likely
+        // to be near 1 if context.vibemodes.suckup == true)
+        return Math.random();
+    }
+
+    public String getContextBasedExcuse(List<Excuse> excuses, Context context, ScenarioType scenarioType) {
+        if (excuses.isEmpty()) {
+            throw new IllegalArgumentException("Non-empty excuses list expected");
+        }
+        List<String> tags = this.getTags(context, scenarioType);
+        List<Excuse> contextBasedExcuses = filterByTag(excuses, tags, 10);
+
+        List<Excuse> finalExcuses = contextBasedExcuses.stream()
+                .sorted(Excuse.byLastUsed)
+                .limit(Math.round(contextBasedExcuses.size() / 2.0))
+                .sorted(Excuse.byRating)
+                .limit(Math.round(contextBasedExcuses.size() / 2.0))
+                .collect(Collectors.toList());
+
+        int randomExcuseId = ThreadLocalRandom.current()
+                .nextInt(0, finalExcuses.size());
+
+        return Wildcards.replace(finalExcuses.get(randomExcuseId)
+                .getText(), context.getWildcardData());
+    }
+
+    /**
+     * This method filters the given excuses by the given tags in a way that the
+     * result contains at least n excuses
+     * 
+     * @param excuses
+     *            to be filtered
+     * @param tags
+     *            to be filtered by. The first tag in the list has the highest
+     *            priority and the last one the lowest
+     * @param n
+     *            minimal amount of excuses to be returned
+     * @return filtered excuses
+     */
+    private List<Excuse> filterByTag(List<Excuse> excuses, List<String> tags, int n) {
+        if (tags.isEmpty()) {
+            return excuses;
+        }
+
+        List<Excuse> filteredExcuses = excuses.stream()
+                .filter(excuse -> excuse.getTags()
+                        .contains(tags.remove(0)))
+                .collect(Collectors.toList());
+
+        // make sure that at least n excuses are returned
+        if (filteredExcuses.size() < n) {
+            return excuses;
+        } else {
+            return filterByTag(filteredExcuses, tags, n);
+        }
+    }
+
+    /**
+     * Generates a list of tags based on context and scenario type
+     * 
+     * @return
+     */
+    private List<String> getTags(Context context, ScenarioType scenarioType) {
+        List<String> tags = new ArrayList<>();
+        tags.add(this.getScenarioTypeTag(scenarioType));
+        tags.addAll(this.getExcusesVibeTags(context));

```

Comment:
Warum machen wir das nicht schon beim Erstellen des Kontexts im ContextHandler? Also das wir im Kontext ne Liste von verschiedenen Tags ablegen, statt booleans für jeden einzelnen. So wie es auch bei LecturerTags und UserInterests weiter unten ist. 
Fände das Konstrukt context.getXY schöner als hier in der Methode öfters this.getXY(context) aufzurufen. Oder was waren die Gründe dafür?